design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,290 @@
+REST endpoint `/reportlag`. The lag information at a server is obtained via 
+`Map<String, Map<Integer, Long>> localLagMap = kafkaStreams.allLocalOffsetLags();`.
+
+Pseudocode for REST endpoint for `/reportlag`:
+```java
+Map<KsqlServer, Map<String, Map<Integer, Long>> globalLagMap;
+
+@POST
+public Response receiveLagInfo(Request lagMap) 
+  return Response.ok(processRequest()).build();
+}
+
+private LagReportResponse processRequest() {
+  KsqlServer server = request.getServer();
+  Map<String, Map<Integer, Long>> localLagMap = request.getLag();
+  globalLagMap.put(server, localLagMap);
+}
+``` 
+
+### Lag-aware routing
+
+Given the above information (alive + lag), how does a KSQL server decide to which standby to route 
+the request? Every server knows the lag of all partitions hosted by all servers in the cluster. 
+This allows us to implement lag-aware routing where server `B` can a) determine that server `A` is 
+down and b) decide whether it will serve the request itself or forward it to `C` depending on who 
+has the smallest lag for the given key in the pull query. 
+
+Pseudocode for routing:
+```java
+// Map populated periodically through heartbeat information
+Map<KsqlNode, Boolean> aliveNodes;
+
+// Map populated periodically through lag reporting
+Map<KsqlNode, Map<String, Map<Integer, Long>> globalLagMap;
+ 
+// Fetch the metadata related to the key
+KeyQueryMetadata queryMetadata = queryMetadataForKey(store, key, serializer);
+ 
+// Acceptable lag for the query
+final long acceptableOffsetLag = 10000;
+
+// Ordered list of servers, in order of most caught-up to least
+List<KsqlNode> nodesToQuery;
+KsqlNode active = queryMetadata.getActiveHost();
+if (aliveNodes.get(active)) {
+  nodesToQuery.add(active)  
+}
+
+// add standbys
+nodesToQuery.addAll(queryMetadata.getStandbyHosts().stream()
+    // filter out all the standbys that are down
+    .filter(standbyHost -> aliveNodes.get(standByHost) == null)
+    // get the lag at each standby host for the key's store partition
+    .map(standbyHost -> new Pair(standbyHostInfo, globalLagMap.get(standbyHost).get(storeName).get(queryMetadata.partition())))
+    // Sort by offset lag, i.e smallest lag first
+    .sorted(Comaparator.comparing(Pair::getRight())
+    .filter(standbyHostLagPair -> standbyHostLagPair.getRight() < acceptableOffsetLag)
+    .map(standbyHostLagPair -> standbyHostLagPair.getLeft())
+    .collect(Collectors.toList()));
+ 
+// Query available nodes, starting from active if up
+List<Row> result = null;
+for (KsqlNode server : nodesToQuery) {
+  try {
+    result = query(store, key, server);
+  } catch (Exception e) {
+    System.err.println("Querying server %s failed", server);
+  }
+  if (result != null) {
+    break;
+  }
+}
+
+if (result == null) {
+  throw new Exception("Unable to serve request. All nodes are down or too far behind");
+}
+
+return result;
+```
+
+We will also introduce a configuration `ksql.query.pull.acceptable.offset.lag`, that will provide 
```

Comment:
Makes sense.. It can be a table property (I atleast need to understand how these work today).

Reply:
Makes sense but how would one provide it as a table/query property? I was thinking for this to be a configuration parameter in `ksql-server.properties`. One does not know before hand what tables/queries will be created. So, will it be part of the CTAS command?

Reply:
For table it could be part of the CTAS command's `WITH` clause, or provided in the request JSON (we support providing properties this way in the current API). In the query we could include it in the request JSON.

Reply:
Would we want this supported in the `WITH` clause?  Maybe... but certainly being able to `SET` it before executing a query would be useful.

Reply:
How can you control this config in a per-table basis using `SET` though?

Reply:
Tracked in issue #4226